### PR TITLE
Reconstruction pipeline

### PR DESCRIPTION
**Pull request by adler-j**
TODO:

- [x] [#1184](https://github.com/odlgroup/odl/pull/1184) "TST: Add general tests for supervised FOMs"
- [x] [#1185](https://github.com/odlgroup/odl/pull/1185) "ENH: Pull request of group 1 (Workshop day, work on FOMs)"
- [x] [#1187](https://github.com/odlgroup/odl/pull/1187) "Optimize parameters"
- [x] <s>[#1188](https://github.com/odlgroup/odl/pull/1188)</s> [#1357](https://github.com/odlgroup/odl/pull/1357) "Reco pipeline mob"

After these are in we can review and merge to master.

Edited by @aringh

### CODE REVIEW

**Comment by sbanert** [comment]
This branch should be rebased on master to include the newest travis fixes.

**Comment by aringh** [comment]
Since all 4 PRs going into this branch has been reviewed before, I guess that what is left here I guess is to rebase on master and after that check that all tests work (travis don't run them, they have to be run manually). Anyone want to do the honors?

**Comment by sbanert** [comment]
I’ll do this.

**Comment by sbanert** [comment]
The rebased branch is in my repository, so any owner may take it and push it to odlgroup/odl.
https://github.com/sbanert/odl/tree/reconstruction_pipeline

**Comment by aringh** [comment]
When I try to check out the branch from @sbanert's repo I get into a "detached head" state, from which I do not mange to push the commits onto this branch. Is there a simple way to do this, with minimal risk for screwing things up? Or is it easier to just close this PR and create a new one from @sbanert's branch?

More specifically: `git push origin reconstruction_pipeline` just returns `Everything up-to-date` which should not be true, and `git push origin/reconstruction_pipeline` gave a long message I do not understand on "default behavior changes" and finally
```
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push origin/reconstruction_pipeline HEAD:<name-of-remote-branch>
```
but I do not understand how to use that (and I don't want to create a mess).

**Comment by kohr-h** [comment]
@aringh The full command for `git push` (to the extent that I know it) is
```
git push <remote> <source_branch>:<target_branch>
```
This will push the `<source_branch>` to `<remote>` and create or update the `<target_branch>` accordingly. After that, the `<source_branch>` and `<remote>/<target_branch>` will be even.
(Actually, I believe that `<source_branch>` doesn't even have to be a branch, it can be an arbitrary commit.)

Anything shorter than that makes assumptions:
- If you leave out the `<target_branch>`, git assumes that a branch of the same name should either be created or pushed to.
  That's why something like `git push origin aringh/master` does works in a weird way: it creates a new branch `aringh/master` at `origin`, usually that's not what people want.
- If you also leave out `<source_branch>` and `<remote>`, then git assumes that (1) you are currently on a local branch and (2) that this branch tracks some remote branch.

So your first attempt, `git push origin reconstruction_pipeline` just pushed your local branch to `origin`, and if they're even, there's nothing to do.
Your second attempt, `git push origin/reconstruction_pipeline` didn't specify a remote and failed therefore. (If you give explicit branches, you also need an explicit remote).

Long story short, the correct way to do this is
```
git push origin sbanert/reconstruction_pipeline:reconstruction_pipeline
```
You also don't need to `checkout` the `sbanert` branch, a `git fetch` is enough.

Small aside, the more you use git, the more you start disliking shorthands like `git pull` or `git push` without arguments. I hardly use them anymore, in favor of the slightly longer but more explicit variants.

**Comment by aringh** [comment]
@kohr-h: thank you very much for taking your time to explain this in a very pedagogic way!

**Comment by aringh** [comment]
**Log**: the build fail was due to a failing doc-test for operator norm
```
=================================== FAILURES ===================================
______________ [doctest] odl.operator.oputils.power_method_opnorm ______________
152     est_opnorm : float
153         The estimated operator norm of ``op``.
154 
155     Examples
156     --------
157     Verify that the identity operator has norm 1:
158 
159     >>> space = odl.uniform_discr(0, 1, 5)
160     >>> id = odl.IdentityOperator(space)
161     >>> power_method_opnorm(id)
Expected:
    1.0
Got:
    1.0000000000000002
```
This should not have been changed in any of these commits, but should potentially be addressed elsewhere. I will rerun travis to see if it is an error due to bad random initialization of the power method.

**Comment by aringh** [comment]
Also, this should be ready for a final review if that is desired. Note however that all PRs going in to form this PR have been review, so I guess it should be more or less ok.

**Comment by sbanert** [comment]
We should make sure that the merging is correct. I also noticed the random failures when I merged everything, but only corrected the non-random ones (the monotonicity test for apparently non-monotonic FOMs). In particular, a review should make sure that the tests which are run for all FOMs do the right thing.

File: odl/contrib/param_opt/param_opt.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+__all__ = ('optimal_parameters', )
+
+
+def optimal_parameters(reconstruction, fom, phantoms, data,
+                       initial_param=0):
+    """Find the optimal parameters for a reconstruction method.
+
+    Notes
+    -----
+    For a forward operator :math:`A : X \\to Y`, a reconstruction operator
```
Comment (adler-j) [inline odl/contrib/param_opt/param_opt.py]: Prefer to change this docstring to use  `r"""` and then remove the double backslashes

File: odl/contrib/param_opt/test/test_param_opt.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def test_optimal_parameters_one_parameter(space, fom):
+    """Tests if optimal_parameters works for some simple examples."""
+    mynoise = [odl.phantom.white_noise(space) for _ in range(2)]
+    phantoms = mynoise.copy()
+    data = mynoise.copy()
+
+    def reconstruction(data, lam):
+        """Perturbs the data by adding lam to it."""
```
Comment (adler-j) [inline odl/contrib/param_opt/test/test_param_opt.py]: This comment is quite un-informative (it's obvious from the code). We should either expand it to add something that is not obvious, or simply remove it.

File: odl/contrib/fom/test/test_supervised.py [hunk 3]
```diff
@@ -60,24 +94,133 @@ def test_mean_absolute_error(space):
...
+    assert pytest.approx(fom.mean_value_difference(I0, I0)) == 0
+    assert fom.mean_value_difference(10 * I0, I0, normalized=True) <= 1.0
+
+
+def test_standard_deviation_difference_range_value(space):
+    I0 = space.element(np.random.normal(0, 1, size=space.shape))
+    const = np.random.normal(0, 10)
+
+    assert pytest.approx(fom.standard_deviation_difference(I0, I0)) == 0
+    assert (fom.standard_deviation_difference(10 * I0, I0, normalized=True) <=
```
[the remark this answers is not in the document]
Comment (sbanert) [reply odl/contrib/fom/test/test_supervised.py]: Because of the `normalized` flag.

**Comment by sbanert** [comment]
Fixes for @adler-j ’s comments up to and including https://github.com/odlgroup/odl/pull/1280#discussion_r194685040 can be found in my repository under the branch `reconstruction_pipeline`: https://github.com/sbanert/odl/tree/reconstruction_pipeline.

**Comment by kohr-h** [comment]
> Fixes for @adler-j ’s comments up to and including #1280 (comment) can be found in my repository under the branch reconstruction_pipeline: https://github.com/sbanert/odl/tree/reconstruction_pipeline.

This workflow is quite suboptimal, mostly because the PR is in the main repo. We should probably avoid that in the future and instead pull from each other in the collaborative phase, and then decide who makes the final PR. I'm sure there are ways to give access to others for selected branches as well.

In the meanwhile, @sbanert, maybe just make a PR into this branch and we'll merge it right away. That's easier than somebody manually pushing from your repo.

**Comment by aringh** [comment]
I agree with @khor-h. However, the branches seems to be a bit out-of-date with each other so I just `cherry-pick`ed the commits this time.

**Comment by aringh** [comment]
I think all comments are addressed. Another round of review, or are we ready to merge?

**Comment by sbanert** [comment]
Don’t merge. After resolving the conflicts, I forgot to run the tests in `param_opt`, and on Wednesday, when working together wtih @zickert, we noticed that some are failing due to changes in the interface of scipy and pytest. I’ll fix this tomorrow.

**Comment by aringh** [comment]
I don't get test fails... I ran all tests manually and everything passed (20 times in a row, to reduce risk for random test fails). I have the following versions of the packages
```
numpy                     1.12.1             py36_nomkl_0  [nomkl]
pytest                    3.0.5                    py36_0  
pytest-cache              1.0                      py36_0  
pytest-pep8               1.0.6                    py36_0
scipy                     0.19.0          np112py36_nomkl_0  [nomkl]

```

**Comment by sbanert** [comment]
I’ve got `scipy 1.1.0`, and this requires to import `scipy.optimize` instead of only `scipy`. On my machine, I could not reproduce the assertion errors I got on shannon.

**Comment by sbanert** [comment]
I get some assertion errors like
```
E       assert [3.5551206290140457e-11, 5.2841825858110269e-10] == approx([0 ± 1.0e-04, 0 ± 1.0e-04])
E        +  where [3.5551206290140457e-11, 5.2841825858110269e-10] = list(array([  3.55512063e-11,   5.28418259e-10]))
E        +  and   approx([0 ± 1.0e-04, 0 ± 1.0e-04]) = <function approx at 0x7f1046335a60>([0, 0], abs=0.0001)
E        +    where <function approx at 0x7f1046335a60> = pytest.approx
```
(which looks as if `numpy` and `pytest.approx` are not working well together) with the configuration
```
numpy                     1.12.1           py36he24570b_1
pytest                    3.6.0                    py36_1  
pytest-arraydiff          0.2                      py36_0  
pytest-astropy            0.4.0                    py36_0  
pytest-cache              1.0                      py36_0  
pytest-doctestplus        0.1.3                    py36_0  
pytest-openfiles          0.3.0                    py36_0  
pytest-pep8               1.0.6                    py36_0  
pytest-remotedata         0.3.0                    py36_0  
python                    3.6.5                hc3d631a_2  
scipy                     1.1.0            py36hfc37229_0  
```

**Comment by aringh** [comment]
I cannot reproduce the errors by @sbanert... I have tried to create an environment with the same packages and run the tests (for foms and param_opt), and I do not get this kind of errors.

```
conda create -n test_env python=3.6
source activate test_env
conda install numpy=1.12.1
conda install -c anaconda python=3.6.5
conda install -c anaconda pytest=3.6.0
conda install -c anaconda scipy=1.1.0
pip install --editable .
```
The last line is installing odl. Not however, that this gave me a numpy version 1.14.5, and I do not seem to be able to downgrade to 1.12.1 due to conflicts...

Result of running the tests = No errors

```
conda install -c anaconda pytest-arraydiff=0.2
```

Results of running tests = No errors

```
conda install -c anaconda pytest-astropy=0.4.0 pytest-cache=1.0 pytest-doctestplus=0.1.3 pytest-openfiles=0.3.0 pytest-pep8=1.0.6 pytest-remotedata=0.3.0
```

Results of running tests  = No errors.

**Comment by aringh** [comment]
Should we merge anyway, or try to locate and fix the issue?

@sbanert could you create a new environment and see if the error still prevails? What happens if you update to numpy 1.14.5? (Although I am pretty sure that I run 1.12.1 in other conda environments...)

**Comment by sbanert** [comment]
If I execute the commands you gave in your last post, I end up with numpy 1.12.1, and I still get these failures. Updating to 1.14.5 solves this, and all the tests pass.

**Comment by aringh** [comment]
Good that it works for 1.14.5. Then maybe we can merge now? :smiley: 

Also: that is very weird... I get the following
```
conda install -c anaconda scipy=1.1.0
Fetching package metadata ...........
Solving package specifications: .

Package plan for installation in environment /home/axel/miniconda3/envs/test_env:

The following NEW packages will be INSTALLED:

    blas:           1.0-mkl               anaconda
    intel-openmp:   2018.0.3-0            anaconda
    libgfortran-ng: 7.2.0-hdf63c60_3      anaconda
    mkl_fft:        1.0.1-py36h3010b51_0  anaconda
    mkl_random:     1.0.1-py36h629b387_0  anaconda
    numpy-base:     1.14.5-py36hdbf6ddf_0 anaconda
    scipy:          1.1.0-py36hfc37229_0  anaconda

The following packages will be UPDATED:

    mkl:            2017.0.3-0                     --> 2018.0.3-1            anaconda
    numpy:          1.12.1-py36_0                  --> 1.14.5-py36hcd700cb_0 anaconda

Proceed ([y]/n)? y
```
which force-updates my numpy... on the other hand, trying to do downgrade it I get the following error

```
conda install numpy=1.12.1
Fetching package metadata .........
Solving package specifications: .


UnsatisfiableError: The following specifications were found to be in conflict:
  - numpy 1.12.1*
Use "conda info <package>" to see the dependencies for each package.
```

**Comment by aringh** [comment]
I feel ready to merge this. Anyone have more comments on this?

**Comment by adler-j** [comment]
I'll have to do another full review. Give me a bit

**Comment by kohr-h** [comment]
Hey, this is currently the last blocker for [Release 0.7](https://github.com/odlgroup/odl/projects/4). Any chance to get it done soon?

**Comment by aringh** [comment]
> Hey, this is currently the last blocker for Release 0.7. Any chance to get it done soon?

I know :wink: That's actually why I was somewhat eager to get it in as well

**Comment by adler-j** [comment]
Its my fault, I'll review

File: odl/contrib/fom/supervised.py [hunk 4]
```diff
@@ -90,9 +102,16 @@ def mean_absolute_error(data, ground_truth, mask=None, normalized=False):
     ground_truth : `Tensor`
         Reference to compare ``data`` to.
     mask : `array-like`, optional
-        If given, ``data * mask`` is compared to ``ground_truth * mask``.
+        If given, ``data * mask`` is compared to ``ground_truth * mask``,
```
Comment (adler-j) [inline odl/contrib/fom/supervised.py]: This is importantly different from `data[mask]` vs `ground_truth[mask]` (this is biased towards 0), perhaps it should be mentioned

File: odl/contrib/fom/supervised.py [hunk 5]
```diff
@@ -193,8 +222,9 @@ def mean_value_difference(data, ground_truth, mask=None, normalized=False):
 
 
 def standard_deviation_difference(data, ground_truth, mask=None,
-                                  normalized=False):
-    """Return absolute difference in std between ``data`` and ``ground_truth``.
+                                  normalized=False,
+                                  force_lower_is_better=False):
+    r"""Return absolute difference in std between ``data`` and ``ground_truth``.
```
Comment (adler-j) [inline odl/contrib/fom/supervised.py]: I think we decided to not have "Return ..." at the start of these strings

File: odl/contrib/fom/supervised.py [hunk 6]
```diff
@@ -328,7 +369,7 @@ def range_difference(data, ground_truth, mask=None, normalized=False):
 
 def blurring(data, ground_truth, mask=None, normalized=False,
              smoothness_factor=None):
-    """Return weighted L2 distance, emphasizing regions defined by ``mask``.
+    r"""Return weighted L2 distance, emphasizing regions defined by ``mask``.
```
Comment (adler-j) [inline odl/contrib/fom/supervised.py]: How is this different from the L2 distance above?

File: odl/contrib/fom/supervised.py [hunk 7]
```diff
@@ -396,71 +437,60 @@ def blurring(data, ground_truth, mask=None, normalized=False,
...
 
 
-def false_structures(data, ground_truth, mask=None, normalized=False,
-                     smoothness_factor=None):
-    """Return weighted L2 distance, de-emphasizing regions defined by ``mask``.
-
-    .. note::
-        If the mask argument is omitted, this FOM is equivalent to the
-        mean squared error.
+def false_structures_mask(foreground, smoothness_factor=None):
```
Comment (adler-j) [inline odl/contrib/fom/supervised.py]: IMO this name makes about 0 sense to what the function does. Either emphasize it in the doc, or rename.

I mean, what is a false structure?

File: odl/contrib/fom/supervised.py [hunk 8]
```diff
@@ -396,71 +437,60 @@ def blurring(data, ground_truth, mask=None, normalized=False,
...
     """
     from scipy.ndimage.morphology import distance_transform_edt
+    space = foreground.space
 
-    if smoothness_factor is None:
-        smoothness_factor = np.mean(data.space.shape) / 10
+    unique = np.unique(foreground)
+    if not np.array_equiv(unique, [0., 1.]):
+        raise ValueError('`foreground` is not a binary mask or has '
+                         'either only true or only false values {!r}'
```
Comment (adler-j) [inline odl/contrib/fom/supervised.py]: Why is this a problem? (all true/false)

File: odl/contrib/fom/test/test_supervised.py [hunk 9]
```diff
@@ -6,24 +6,58 @@
...
                            odl.util.testutils.skip_if_no_pyfftw('pyfftw')])
+
 space = simple_fixture('space',
                        [odl.rn(3),
-                        odl.uniform_discr(0, 1, 10)])
+                        odl.rn(3, dtype='float32'),
+                        odl.uniform_discr(0, 1, 10),
+                        odl.uniform_discr([0, 0], [1, 1], [5, 5])])
+
+scalar_fom = simple_fixture('scalar_fom',
```
Comment (adler-j) [inline odl/contrib/fom/test/test_supervised.py]: Lots of other FOMS are scalars (e.g. PSNR), is there some other feature that's important to these? Name should reflect this

File: odl/contrib/fom/test/test_supervised.py [hunk 10]
```diff
@@ -6,24 +6,58 @@
...
 import scipy.signal
 import scipy.misc
 import odl
-import odl.contrib.fom
-from odl.contrib.fom.util import filter_image_sep2d
-from odl.util.testutils import simple_fixture
+from odl.contrib import fom
+from odl.util.testutils import simple_fixture, noise_element
 
 fft_impl = simple_fixture('fft_impl',
```
Comment (adler-j) [inline odl/contrib/fom/test/test_supervised.py]: Don't we have a global fixture for this @kohr-h?

File: odl/contrib/fom/test/test_supervised.py [hunk 11]
```diff
@@ -60,23 +94,135 @@ def test_mean_absolute_error(space):
...
+    # Test with Z-score that result is independent of affine transformation
+    result = fom.psnr(data * 3.7 + 1.234, true, use_zscore=True)
+    expected = fom.psnr(data, true, use_zscore=True)
+    assert result == pytest.approx(expected, abs=1e-5)
+
+
+def test_ssim(space):
+    ground_truth = odl.phantom.white_noise(space)
+
+    # SSIM of true image should be either
```
Comment (adler-j) [inline odl/contrib/fom/test/test_supervised.py]: SSIM of true image compared to ?

File: odl/contrib/param_opt/param_opt.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+__all__ = ('optimal_parameters', )
+
+
+def optimal_parameters(reconstruction, fom, phantoms, data,
+                       initial_param=0):
+    r"""Find the optimal parameters for a reconstruction method.
+
+    Notes
+    -----
+    For a forward operator :math:`A : X \to Y`, a reconstruction operator
```
Comment (adler-j) [inline odl/contrib/param_opt/param_opt.py]: This function doesn't even seem to care about forward operators at all. It's all inference.

Are we sure we want to write this using forward operators?

File: odl/contrib/param_opt/test/test_param_opt.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+                        odl.uniform_discr(0, 1, 10)])
+
+fom = simple_fixture('fom',
+                     [odl.contrib.fom.mean_squared_error,
+                      odl.contrib.fom.mean_absolute_error])
+
+
+def test_optimal_parameters_one_parameter(space, fom):
+    """Tests if optimal_parameters works for some simple examples."""
+    mynoise = [odl.phantom.white_noise(space) for _ in range(2)]
```
Comment (adler-j) [inline odl/contrib/param_opt/test/test_param_opt.py]: We don't use "my..." in actual code, it looks like a code example :)

File: odl/contrib/param_opt/test/test_param_opt.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        return data + lam
+
+    result = odl.contrib.param_opt.optimal_parameters(reconstruction, fom,
+                                                      phantoms, data, 1)
+    assert result == pytest.approx(0, abs=1e-4)
+
+
+def test_optimal_parameters_two_parameters(space, fom):
+    """Tests if optimal_parameters works for some simple examples."""
+    mynoise = [odl.phantom.white_noise(space) for _ in range(2)]
```
Comment (adler-j) [inline odl/contrib/param_opt/test/test_param_opt.py]: Se above on "my..."

File: odl/contrib/param_opt/test/test_param_opt.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    phantoms = mynoise.copy()
+    data = mynoise.copy()
+
+    def reconstruction1(data, params):
+        """Perturbs the data by adding the sum of squares of params to it."""
+        return data + sum([param ** 2 for param in params])
+
+    result1 = odl.contrib.param_opt.optimal_parameters(reconstruction1, fom,
+                                                       phantoms, data, [1, 2])
+    assert list(result1) == pytest.approx([0, 0], abs=1e-4)
```
Comment (adler-j) [inline odl/contrib/param_opt/test/test_param_opt.py]: Any chance for examples that don't have 0 as their optimum? I'd rather have something weird (e.g. 0.8234) so that we don't happen to end up there by pure luck.

File: odl/contrib/fom/test/test_supervised.py (hunk 10 above)
Comment (kohr-h) [reply odl/contrib/fom/test/test_supervised.py]: Not global AFAIK

File: odl/contrib/fom/supervised.py (hunk 6 above)
Comment (aringh) [reply odl/contrib/fom/supervised.py]: It looks like it does some kind of smoothing before it takes an L2 distance... I don't know. Maybe we can just add that possibility to the mean squared error function instead?

File: odl/contrib/fom/supervised.py (hunk 7 above)
Comment (aringh) [reply odl/contrib/fom/supervised.py]: `foreground_mask`?

File: odl/contrib/fom/supervised.py (hunk 8 above)
Comment (aringh) [reply odl/contrib/fom/supervised.py]: I guess it does not make sense, from some perspective? If everything is in the foreground, then no mask is needed. If everything is background, we have nothing to compare to. But no, technically it might not be ok.

Comment (adler-j) [reply odl/contrib/fom/supervised.py]: If it is technically OK we should allow it, else you get irritating bugs. e.g. what happens if you loop over a list of `foregrounds=[nothing, some stuff, everything]`, currently that would (irritatingly) give an error

File: odl/contrib/fom/supervised.py (hunk 6 above)
Comment (adler-j) [reply odl/contrib/fom/supervised.py]: L2 already has a mask, I don't see any difference.

Comment (aringh) [reply odl/contrib/fom/supervised.py]: The return statement in the function is `mean_squared_error(data, ground_truth, mask, normalized)`, and the only thing it does before is some preprocessing of the mask. So yes, it is obvious that this is the L2 distance.

However, the real question should maybe be *why* someone thought that this kind of preprocessing of the mask was relevant. That I cannot answer this, but I assume that it does some blurring to, e.g., reduce the effect to off-by-one-pixel misalignment in a reconstruction. If someone knows what the idea behind this was, please fill in. @davlars @zickert @ozanoktem @bgris or anyone else how could have been here when it was first implemented.

My suggestion: redo it as a utility function that returns the smoothed mask, which can then be applied in any fom that takes a mask as input. Sounds ok?

File: odl/contrib/fom/test/test_supervised.py (hunk 11 above)
Comment (aringh) [reply odl/contrib/fom/test/test_supervised.py]: `SSIM of an image compared to itself should be either`

File: odl/contrib/param_opt/param_opt.py (hunk 12 above)
Comment (aringh) [reply odl/contrib/param_opt/param_opt.py]: We can rewrite. But in that case we should maybe rename most of the parameters as well? Their names are somehow also taken from what they represent in an inverse problem.

(`reconstruction` -> `operator`, `phantom` -> `phantom` or `ground_truth`, `data` -> ???)

File: odl/contrib/param_opt/test/test_param_opt.py (hunk 15 above)
Comment (aringh) [reply odl/contrib/param_opt/test/test_param_opt.py]: I hope the updated version fulfill your wish :wink:

File: odl/contrib/fom/supervised.py (hunk 8 above)
Comment (kohr-h) [reply odl/contrib/fom/supervised.py]: Agreed, even if for no other reason but performance (we do `np.unique` just for the check)

File: odl/contrib/param_opt/param_opt.py (hunk 12 above)
Comment (adler-j) [reply odl/contrib/param_opt/param_opt.py]: This became complicated, lets leave it :D

File: odl/contrib/param_opt/param_opt.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+                   for phantomi, datai in zip(phantoms, data))
+
+    # Pick resolution to fit the one used by the space
+    tol = np.finfo(phantoms[0].space.dtype).resolution * 10
+
+    initial_param = np.asarray(initial_param)
+
+    # We use a faster optimizer for the one parameter case
+    if initial_param.size == 1:
+        bracket = [initial_param - tol, initial_param + tol]
```
Comment (olivierverdier) [inline odl/contrib/param_opt/param_opt.py]: Since `tol` is typically ridiculously small (1e-14), won't this bracket also be ridiculously narrow?

Comment (adler-j) [reply odl/contrib/param_opt/param_opt.py]: Sure will, but that's fine if we read the doc:

```
bracket : sequence, optional
    For methods 'brent' and 'golden', `bracket` defines the bracketing
    interval and can either have three items ``(a, b, c)`` so that
    ``a < b < c`` and ``fun(b) < fun(a), fun(c)`` or two items ``a`` and
    ``c`` which are assumed to be a starting interval for a downhill
    bracket search (see `bracket`); it doesn't always mean that the
    obtained solution will satisfy ``a <= x <= c``.
```

With that said, users should probably not need to read the doc, and this should me mentioned in a comment.

Comment (kohr-h) [reply odl/contrib/param_opt/param_opt.py]: There's also a single backtick issue with \`bracket\` while we're at it. And `'brent'` and `'golden'` should be double-backticked as well.

Comment (adler-j) [reply odl/contrib/param_opt/param_opt.py]: This is litterally copy paste from scipy :D

Comment (olivierverdier) [reply odl/contrib/param_opt/param_opt.py]: You are right, the bracket could be anything, and that bracket, or the default one, works for some nice functions. However, I did some tests on Shepp-Logan+FPB(cutoff)+SSIM, and, in that case, that tiny bracket really seems to be a problem: the minimiser always returns the initial value. Note that the default bracket does not work either (but the result is different). The function to minimise looks like ~the Texas border~ [that](https://imgur.com/nRUTDtW).

Comment (adler-j) [reply odl/contrib/param_opt/param_opt.py]: Perhaps we should allow `initial_param` to be either a point value or a bracket then!

File: odl/contrib/fom/supervised.py (hunk 6 above)
Comment (zickert) [reply odl/contrib/fom/supervised.py]: Regarding why this was implemented at the first "reconstruction pipeline workshop", I think inspiration for some of the figures of merit (including this one) were taken from section 2.4 in this paper: https://www.sciencedirect.com/science/article/pii/S1047847701943382

@aringh's suggestion sounds reasonable to me.

**Comment by kohr-h** [comment]
How are we doing on this one? C'mon people, this is the last item blocking the 0.7 release #1335 :wink:

**Comment by adler-j** [comment]
I agree with @kohr-h, this really has to go in. I'll do this myself to get it out of the way.

**Comment by kohr-h** [comment]
> I agree with @kohr-h, this really has to go in. I'll do this myself to get it out of the way.

:+1: I'll volunteer to make the release, during the weekend if possible.

File: odl/contrib/param_opt/param_opt.py (hunk 16 above)
Comment (adler-j) [reply odl/contrib/param_opt/param_opt.py]: Ignoring this for now, made an issue: 1422

File: odl/contrib/param_opt/test/test_param_opt.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    assert result == pytest.approx(1.23, abs=1e-4)
+
+
+def test_optimal_parameters_two_parameters(space, fom):
+    """Tests if optimal_parameters works for some simple examples."""
+    noise = [odl.phantom.white_noise(space) for _ in range(2)]
+    # Normalize to reduce test fails due to randomness
+    noise = [noise_elem / noise_elem.norm() for noise_elem in noise]
+    phantoms = noise.copy()
+    data = noise.copy()
```
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/param_opt/test/test_param_opt.py]: Easier to read further down

**Comment by adler-j** [comment]
Updated according to comments except one, which I made a new issue for. Final review if needed, else merge.

**Comment by ozanoktem** [comment]
Many of the more unusual FOMs are defined in https://www.sciencedirect.com/science/article/pii/S1047847701943382

File: odl/contrib/param_opt/param_opt.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        result = scipy.optimize.minimize_scalar(func,
+                                                bracket=bracket,
+                                                tol=tol,
+                                                bounds=None,
+                                                options={'disp': False})
```
Comment (kohr-h) [inline odl/contrib/param_opt/param_opt.py]: This stacking of arguments on the right looks... not great. Please break after the first opening paren instead and use the second line more.

File: odl/contrib/param_opt/param_opt.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        return result.x
+    else:
+        # Use a gradient free method to find the best parameters
+        parameters = scipy.optimize.fmin_powell(func, initial_param,
+                                                xtol=tol,
+                                                ftol=tol,
+                                                disp=False)
```
Comment (kohr-h) [inline odl/contrib/param_opt/param_opt.py]: Same here

File: odl/contrib/fom/supervised.py [hunk 20]
```diff
@@ -15,12 +15,13 @@
...
-           'range_difference', 'blurring', 'false_structures', 'ssim', 'psnr',
-           'haarpsi', 'noise_power_spectrum')
+           'range_difference', 'blurring', 'false_structures_mask', 'ssim',
+           'psnr', 'haarpsi', 'noise_power_spectrum')
 
 
-def mean_squared_error(data, ground_truth, mask=None, normalized=False):
-    """Return mean squared L2 distance between ``data`` and ``ground_truth``.
+def mean_squared_error(data, ground_truth, mask=None,
+                       normalized=False, force_lower_is_better=False):
```
[the remark this answers is not in the document]
Comment (sbanert) [reply odl/contrib/fom/supervised.py]: I think that the default value for this flag should always be false because that’s what one usually would expect when calling a function like SSIM where `force_lower_is_better=True` would reverse the sign.

[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/supervised.py]: Sure, but why shouldn't different FOMs have different defaults?

[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/supervised.py]: Actually, the whole reason for this oddity is the prefix `force_`. It makes the parameter a flag for *doing something* vs *doing nothing*. This is a much narrower construct than a flag for *doing one thing* vs *doing another thing*.
If the parameter was called `lower_is_better`, it would be quite natural to set its default value to `False` for SSIM and other measures that are *scores*, and to set it to `True` for *error measures*. And still you could enforce uniform behavior by choosing one of them across all FOMs.

[the remark this answers is not in the document]
Comment (sbanert) [reply odl/contrib/fom/supervised.py]: This flag is meant to *change* the default behaviour, but perhaps this is only a matter of proper documentation (which is really an issue: http://odlgroup.github.io/odl/odl.contrib.html). Giving a different default behaviour could (in my opinion) suggest setting this flag changes anything for FOMS which have lower values for better quality.

[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/contrib/fom/supervised.py]: > This flag is meant to change the default behaviour

Yes, and that's a one-way operation currently because there's only one direction into which to change. With my suggestion, the default value *indicates* the default behavior, and you could change it in either direction. IMO that would be a much more intuitive interface.



> Giving a different default behaviour could (in my opinion) suggest setting this flag changes anything for FOMs which have lower values for better quality.

Yes, and I think it should.

File: odl/contrib/fom/supervised.py [hunk 21]
```diff
@@ -328,7 +366,7 @@ def range_difference(data, ground_truth, mask=None, normalized=False):
```
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/fom/supervised.py]: I'll leave this as is for readability. If this ever becomes e.g. a performance bottleneck we can fix.

File: odl/contrib/fom/supervised.py [hunk 22]
```diff
@@ -129,11 +147,16 @@ def mean_absolute_error(data, ground_truth, mask=None, normalized=False):
...
+    # The mean absolute error is already in the desired order.
+    # The output does not need to be changed depending on the
+    # force_lower_is_better flag.
+
     return fom
 
 
-def mean_value_difference(data, ground_truth, mask=None, normalized=False):
-    """Return difference in mean value between ``data`` and ``ground_truth``.
+def mean_value_difference(data, ground_truth, mask=None, normalized=False,
```
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/fom/supervised.py]: Frankly I'll leave these as is right now to get this out of the way. In general I'm very skeptical to even having this function in ODL since it really makes no sense.

File: odl/contrib/fom/supervised.py [hunk 23]
```diff
@@ -249,18 +281,22 @@ def standard_deviation_difference(data, ground_truth, mask=None,
...
+        fom /= sum_deviation
+        if np.isnan(fom):
+            fom = 0
 
     return fom
 
 
-def range_difference(data, ground_truth, mask=None, normalized=False):
-    """Return dynamic range difference between ``data`` and ``ground_truth``.
+def range_difference(data, ground_truth, mask=None, normalized=False,
```
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/fom/supervised.py]: Same as above applies here. I'll leave this in, but we should discuss what these are actually for.

**Comment by adler-j** [comment]
Fixed some of the requested issues and created #1423 where we can discuss the rest. How about we go now?

File: odl/contrib/fom/supervised.py [hunk 24]
```diff
@@ -193,8 +220,9 @@ def mean_value_difference(data, ground_truth, mask=None, normalized=False):
 
 
 def standard_deviation_difference(data, ground_truth, mask=None,
```
[the remark this answers is not in the document]
Comment (adler-j) [reply odl/contrib/fom/supervised.py]: #1423

**Comment by kohr-h** [comment]
> Fixed some of the requested issues and created #1423 where we can discuss the rest. How about we go now?

Yeah let's go ahead. I guess since it's `contrib` after all, we have a better handle on fixing stuff afterwards.
